### PR TITLE
Precision plugin recipes

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by maanug-nv** [comment]
Defaults are taken from [this](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/training/arguments.py#L300-L321) setup in Megatron-LM, and common NeMo 1.0 defaults.

**Comment by maanug-nv** [comment]
depends on https://github.com/NVIDIA/NeMo/pull/10129

**Review by akoumpa — APPROVED** [review]
LGTM, thanks.

File: nemo/collections/llm/recipes/precision/mixed_precision.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        MegatronMixedPrecision,
+        precision="16-mixed",
+        params_dtype=torch.half,
+        pipeline_dtype=torch.half,
+        autocast_enabled=False,
+        grad_reduce_in_fp32=False,
+    )
+
+
+def attach_fp8_to_plugin(cfg: Config[MegatronMixedPrecision]) -> Config[MegatronMixedPrecision]:
```
Comment (hemildesai) [inline nemo/collections/llm/recipes/precision/mixed_precision.py:29]: Can we call this `fp8_plugin` to be consistent? You can call `bf16` or `fp16` inside it then edit the config.

**Review by ShriyaPalsamudram — DISMISSED** [review]

**Review by hemildesai — DISMISSED** [review]

File: nemo/collections/llm/recipes/precision/mixed_precision.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        MegatronMixedPrecision,
+        precision="16-mixed",
+        params_dtype=torch.half,
+        pipeline_dtype=torch.half,
+        autocast_enabled=False,
+        grad_reduce_in_fp32=False,
+    )
+
+
+def bf16_with_fp8_mixed_plugin() -> Config[MegatronMixedPrecision]:
```
Comment (ericharper) [inline nemo/collections/llm/recipes/precision/mixed_precision.py:29]: let's remove fp8 until more verification is done

File: nemo/collections/llm/recipes/precision/mixed_precision.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    cfg = bf16_mixed_plugin()
+    cfg.fp8 = 'hybrid'
+    cfg.fp8_margin = 0
+    cfg.fp8_interval = 1
+    cfg.fp8_amax_history_len = 1024
+    cfg.fp8_amax_compute_algo = "max"
+    return cfg
+
+
+def fp16_with_fp8_mixed_plugin() -> Config[MegatronMixedPrecision]:
```
Comment (ericharper) [inline nemo/collections/llm/recipes/precision/mixed_precision.py:39]: let's remove fp8 until more verification is done

Comment (maanug-nv) [reply nemo/collections/llm/recipes/precision/mixed_precision.py]: moved FP8 to https://github.com/NVIDIA/NeMo/pull/10208

**Review by ShriyaPalsamudram — APPROVED** [review]
LGTM

**Review by ShriyaPalsamudram — APPROVED** [review]

**Pull request merged**